File: GoL/Model/GameLoop.m
//
//  GameLoop.m
//  GoL
//
//  Created by Georg Meyer on 03.05.14.
//
//

#import "GameLoop.h"
#import "Game.h"

@interface GameLoop ()

@property (nonatomic) float timeInterval;
@property (strong, nonatomic) Game *game;
@property (nonatomic) BOOL keepLooping;
@property (nonatomic) BOOL killLoop;

@property (nonatomic, assign) void (^afterKill)();
@property (strong, nonatomic) dispatch_queue_t gameLoopQueue;
@end

@implementation GameLoop

- (id)initWithGame:(Game *)game andTimeInterval:(float)timeInterval andAfterTick:(void (^)()) afterTick {
    self = [super init];
    if (self) {
        self.game = game;
        self.timeInterval = timeInterval;
        self.keepLooping = NO;
        self.killLoop = NO;
        self.afterTick = afterTick;
        self.afterKill = nil;
        self.gameLoopQueue = dispatch_queue_create("Game loop", NULL);

        [self dispatchAsyncLoop];
    }
    return self;
}

- (void)dispatchAsyncLoop {
    void (^afterTick)() = self.afterTick;
    Game *game = self.game;
    GameLoop *gameLoop = self;
    self.keepLooping = YES;
    BOOL *keepLooping = &_keepLooping;
    BOOL *killLoop = &_killLoop;

    dispatch_async(self.gameLoopQueue, ^{
        while (true) {
            if (*keepLooping && !*killLoop) {
                [game tick];
                if (afterTick) {
                    afterTick();
                }
            }
            if(*killLoop) {
                NSLog(@"Kill game loop");
                dispatch_async(dispatch_get_main_queue(), ^{
                    gameLoop.afterKill();
                });

                break;
            }
            [NSThread sleepForTimeInterval:self.timeInterval];
        }
    });
}

- (void)start {
    self.keepLooping = YES;
    NSLog(@"Start game loop");
}

- (void)stop {
    self.keepLooping = NO;
    NSLog(@"Stop game loop");
}

- (void)killWithCallback:(void (^)()) afterKill {
    self.afterKill = afterKill;
    [self stop];
    self.killLoop = YES;
}

- (void)callAfterKill {
    self.afterKill();
}

@end

File: GoL/Model/GameLoop.swift
//
//  GameLoop.swift
//  GoL
//
//  Created by Georg Meyer on 08.06.14.
//
//

import Foundation


class GameLoop {
    var timeInterval : Double
    var game : Game
    var keepLooping : Bool
    var afterTick : () -> ()?
    
    init(game : Game, timeInterval : Double, afterTick : () -> ()){
        self.game = game
        self.timeInterval = timeInterval
        self.keepLooping = false
        self.afterTick = afterTick
        dispatchAsyncLoop()
    }
    
    func dispatchAsyncLoop() {
        let gameLoop = self
        
        let gameLoopQueue = dispatch_queue_create("Game Loop", nil)
        
        dispatch_async(gameLoopQueue, {
            while(true){
                if(gameLoop.keepLooping){
                    gameLoop.game.tick()
                    if(gameLoop.afterTick != nil){
                        gameLoop.afterTick()
                    }
                }
                
                NSThread.sleepForTimeInterval(gameLoop.timeInterval)
            }
        })
    }
    
    func start() {
        keepLooping = true
    }
    
    func stop() {
        keepLooping = false
    }
}